File: app/src/main/java/com/twenk11k/sideproject/volumecontrol/MainActivity.java
package com.twenk11k.sideproject.volumecontrol;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.media.AudioManager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.SeekBar;

import com.twenk11k.sideproject.volumecontrol.listener.OnAudioVolumeChangedListener;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity implements OnAudioVolumeChangedListener, View.OnClickListener {

    private SeekBar seekBarVolume = null;
    private AudioManager audioManager = null;
    private int sbarProgress = 0;
    private AudioVolumeObserver mAudioVolumeObserver;
    private Drawable musicNoteD, musicOffD;
    // Butterknife bindings..
    @BindView(R.id.music_note)
    ImageView musicNote;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        setVolumeControlStream(AudioManager.STREAM_MUSIC);
        setSeekBar();
        musicNoteD = getResources().getDrawable(R.drawable.baseline_music_note_white_24);
        musicOffD = getResources().getDrawable(R.drawable.baseline_music_off_white_24);
        musicNote.setOnClickListener(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mAudioVolumeObserver != null) {
            mAudioVolumeObserver.unregister();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (seekBarVolume != null) {
            if (audioManager != null) {
                int currentVol = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
                seekBarVolume.setProgress(currentVol);
                handleIcon(currentVol);
            }
        }
        if (mAudioVolumeObserver == null) {
            mAudioVolumeObserver = new AudioVolumeObserver(this);
        }
        mAudioVolumeObserver.register(AudioManager.STREAM_MUSIC, this);

    }

    private void setSeekBar() {
        try {
            seekBarVolume = (SeekBar) findViewById(R.id.seekbar);
            seekBarVolume.setSecondaryProgress(100);
            audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
            seekBarVolume.setMax(audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC));
            int volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
            seekBarVolume.setProgress(volume);
            seekBarVolume.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                @Override
                public void onStopTrackingTouch(SeekBar arg0) {
                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,
                            sbarProgress, AudioManager.FLAG_PLAY_SOUND);
                }

                @Override
                public void onStartTrackingTouch(SeekBar arg0) {
                }

                @Override
                public void onProgressChanged(SeekBar arg0, int progress, boolean arg2) {
                    Log.d("progres_changed", String.valueOf(audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)));
                    sbarProgress = progress;
                    handleIcon(progress);
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onAudioVolumeChanged(int currentVolume, int maxVolume) {
        if (seekBarVolume != null) {
            seekBarVolume.setProgress(currentVolume);
        }
        handleIcon(currentVolume);
    }

    private void handleIcon(int val) {
        if (val == 0) {
            musicNote.setImageDrawable(musicOffD);
        } else {
            musicNote.setImageDrawable(musicNoteD);
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.music_note:
                if (musicNote.getDrawable().equals(musicNoteD)) {

                    Utils.setSeekbarProgress(this, sbarProgress);

                    seekBarVolume.setProgress(0);

                    sbarProgress = 0;

                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,
                            sbarProgress, AudioManager.FLAG_PLAY_SOUND);

                    musicNote.setImageDrawable(musicOffD);
                } else {

                    sbarProgress = Utils.getSeekbarProgress(this);
                    Log.d("sbarprogressval",String.valueOf(sbarProgress));
                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,
                            sbarProgress, AudioManager.FLAG_PLAY_SOUND);
                    seekBarVolume.setProgress(sbarProgress);
                    musicNote.setImageDrawable(musicNoteD);
                }
                break;
        }
    }
}

File: app/src/main/java/com/twenk11k/sideproject/volumecontrol/ui/activity/MainActivity.kt
package com.twenk11k.sideproject.volumecontrol.ui.activity

import android.content.Context
import android.graphics.drawable.Drawable
import android.media.AudioManager
import android.os.Bundle
import android.widget.SeekBar
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.twenk11k.sideproject.volumecontrol.R
import com.twenk11k.sideproject.volumecontrol.audiovolume.AudioVolumeObserver
import com.twenk11k.sideproject.volumecontrol.audiovolume.OnAudioVolumeChangedListener
import com.twenk11k.sideproject.volumecontrol.databinding.ActivityMainBinding
import com.twenk11k.sideproject.volumecontrol.pref.ProgressPrefs

class MainActivity : AppCompatActivity(), OnAudioVolumeChangedListener {

    private var audioManager: AudioManager? = null
    private var sBarProgress = 0
    private var audioVolumeObserver: AudioVolumeObserver? = null
    private var musicNoteDrawable: Drawable? = null
    private var musicOffDrawable: Drawable? = null
    private val preferences = ProgressPrefs()

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        volumeControlStream = AudioManager.STREAM_MUSIC
        setViews()
    }

    override fun onResume() {
        super.onResume()
        if (audioManager != null) {
            val currentVol = audioManager!!.getStreamVolume(AudioManager.STREAM_MUSIC)
            binding.seekBar.progress = currentVol
            handleIcon(currentVol)
        }
        if (audioVolumeObserver == null) {
            audioVolumeObserver = AudioVolumeObserver(this)
        }
        audioVolumeObserver?.register(AudioManager.STREAM_MUSIC, this)
    }

    override fun onPause() {
        super.onPause()
        audioVolumeObserver?.unregister()
    }

    private fun setViews() {
        setSeekBar()
        setDrawables()
        setViewListeners()
    }

    private fun setDrawables() {
        musicNoteDrawable = ContextCompat.getDrawable(this, R.drawable.baseline_music_note_white_24)
        musicOffDrawable = ContextCompat.getDrawable(this, R.drawable.baseline_music_off_white_24)
    }

    private fun setViewListeners() {
        binding.imgMusicNote.setOnClickListener {
            if (binding.imgMusicNote.drawable == musicNoteDrawable) {
                preferences.seekBarProgress = sBarProgress
                binding.seekBar.progress = 0
                sBarProgress = 0
                audioManager?.setStreamVolume(
                    AudioManager.STREAM_MUSIC,
                    sBarProgress,
                    AudioManager.FLAG_PLAY_SOUND
                )
                binding.imgMusicNote.setImageDrawable(musicOffDrawable)
            } else {
                sBarProgress = preferences.seekBarProgress
                audioManager?.setStreamVolume(
                    AudioManager.STREAM_MUSIC,
                    sBarProgress,
                    AudioManager.FLAG_PLAY_SOUND
                )
                binding.seekBar.progress = sBarProgress
                binding.imgMusicNote.setImageDrawable(musicNoteDrawable)
            }
        }
    }

    private fun setSeekBar() {
        binding.seekBar.secondaryProgress = 100
        audioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        binding.seekBar.max = audioManager!!.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        val volume = audioManager!!.getStreamVolume(AudioManager.STREAM_MUSIC)
        binding.seekBar.progress = volume
        binding.seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                sBarProgress = progress
                handleIcon(progress)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {}

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                audioManager?.setStreamVolume(
                    AudioManager.STREAM_MUSIC,
                    sBarProgress,
                    AudioManager.FLAG_PLAY_SOUND
                )
            }
        })
    }

    private fun handleIcon(currentVol: Int) {
        if (currentVol == 0) {
            binding.imgMusicNote.setImageDrawable(musicOffDrawable)
        } else {
            binding.imgMusicNote.setImageDrawable(musicNoteDrawable)
        }
    }

    override fun onAudioVolumeChanged(currentVolume: Int, maxVolume: Int) {
        binding.seekBar.progress = currentVolume
        handleIcon(currentVolume)
    }
}
